packages/core/src/browser/shell/tab-bar-toolbar.ts

```diff
@@ -0,0 +1,237 @@
+                    const className = `fa fa-${labelPart.name}${labelPart.animation ? ' fa-' + labelPart.animation : ''}`;
+                    child.classList.add(...className.split(' '));
+                } else {
+                    child.innerText = labelPart;
+                }
+                if (item.tooltip) {
+                    child.title = item.tooltip;
+                }
+                itemContainer.appendChild(child);
+            }
+            this.node.appendChild(itemContainer);
+            this.items.set(item, itemContainer);
+        }
+        if (this.items.size === 0) {
+            this.addClass(TabBarToolbar.Styles.TAB_BAR_TOOLBAR_HIDDEN);
+        } else {
+            this.removeClass(TabBarToolbar.Styles.TAB_BAR_TOOLBAR_HIDDEN);
+        }
+    }
+
+    /**
+     * `true` if `left` and `right` contains the same items in the same order. Otherwise, `false`.
+     * We consider two items the same, if the IDs of the corresponding items are the same.
+     */
+    protected areSame(left: TabBarToolbarItem[], right: TabBarToolbarItem[]): boolean {
+        if (left.length === right.length) {
+            for (let i = 0; i < left.length; i++) {
+                if (left[0].id !== right[0].id) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+}
+
+export namespace TabBarToolbar {
+
+    export namespace Styles {
+
+        export const TAB_BAR_TOOLBAR = 'p-TabBar-toolbar';
+        export const TAB_BAR_TOOLBAR_ITEM = 'item';
+        export const TAB_BAR_TOOLBAR_HIDDEN = 'hidden';
+
+    }
+
+}
+
+/**
+ * Clients should implement this interface if they want to contribute to the tab-bar toolbar.
+ */
+export const TabBarToolbarContribution = Symbol('TabBarToolbarContribution');
+export interface TabBarToolbarContribution {
+
+    registerToolbarItems(registry: TabBarToolbarRegistry): void;
+
+}
+
+/**
+ * Representation of an item in the tab
+ */
+export interface TabBarToolbarItem {
+
+    /**
+     * The unique ID of the toolbar item.
+     */
+    readonly id: string;
+
+    /**
+     * The command to execute.
+     */
+    readonly command: Command;
+
+    /**
+     * Text of the item.
+     *
+     * Shamelessly copied and reused from `status-bar`:
+     *
+     * More details about the available `fontawesome` icons and CSS class names can be hound [here](http://fontawesome.io/icons/).
+     * To set a text with icon use the following pattern in text string:
+     * ```typescript
+     * $(fontawesomeClasssName)
+     * ```
+     *
+     * To use animated icons use the following pattern:
+     * ```typescript
+     * $(fontawesomeClassName~typeOfAnimation)
+     * ````
+     * The type of animation can be either `spin` or `pulse`.
+     * Look [here](http://fontawesome.io/examples/#animated) for more information to animated icons
+     */
+    readonly text: string;
```

Comment:
Support for other icons?

Reply:
Yes, it would be great to make the API independent of fontawesome so we could switch to another icon provider (or implement our own) more easily.

Reply:
Deferred: https://github.com/theia-ide/theia/issues/3395.

Reply:
I propose doing it when it is required and adjust the status bar too. What do you think?